pkg/agent/proxy/proxier_test.go

```diff
@@ -3683,58 +3687,509 @@ func TestServicesWithSameEndpoints(t *testing.T) {
+		Endpoints:   []discovery.Endpoint{},
+		Ports: []discovery.EndpointPort{
+			{
+				Name:     ptr.To("http"),
+				Port:     ptr.To(int32(80)),
+				Protocol: ptr.To(corev1.ProtocolTCP),
+			},
+		},
+	}
+	for _, addr := range addresses {
+		endpointSlice.Endpoints = append(endpointSlice.Endpoints, discovery.Endpoint{Addresses: []string{addr}})
+	}
+	return endpointSlice
+}
+
+func generateSvcAndEps(enableIPv4, enableIPv6 bool) []runtime.Object {
+	objects := make([]runtime.Object, 0, 3)
+
+	var clusterIP string
+	var clusterIPs []string
+	var ipFamilies []corev1.IPFamily
+	if enableIPv4 {
+		clusterIP = "10.96.0.1"
+		clusterIPs = []string{clusterIP}
+		ipFamilies = []corev1.IPFamily{corev1.IPv4Protocol}
+	}
+	if enableIPv6 {
+		if clusterIP == "" {
+			clusterIP = "fd00::1"
+		}
+		clusterIPs = append(clusterIPs, "fd00::1")
+		ipFamilies = append(ipFamilies, corev1.IPv6Protocol)
+	}
+	objects = append(objects, generateSvc(clusterIP, clusterIPs, ipFamilies, 80))
+
+	if enableIPv4 {
+		objects = append(objects, generateEps(discovery.AddressTypeIPv4, []string{"10.244.0.2", "10.244.0.3"}))
+	}
+	if enableIPv6 {
+		objects = append(objects, generateEps(discovery.AddressTypeIPv6, []string{"fd00::100", "fd00::101"}))
+	}
+	return objects
+}
+
+func generateUpdatedSvcAndEps(enableIPv4, enableIPv6 bool) (*corev1.Service, []*discovery.EndpointSlice) {
+	var clusterIP string
+	var clusterIPs []string
+	var ipFamilies []corev1.IPFamily
+	if enableIPv4 {
+		clusterIP = "10.96.0.1"
+		clusterIPs = []string{clusterIP}
+		ipFamilies = []corev1.IPFamily{corev1.IPv4Protocol}
+	}
+	if enableIPv6 {
+		if clusterIP == "" {
+			clusterIP = "fd00::1"
+		}
+		clusterIPs = append(clusterIPs, "fd00::1")
+		ipFamilies = append(ipFamilies, corev1.IPv6Protocol)
+	}
+	svc := generateSvc(clusterIP, clusterIPs, ipFamilies, 8080)
+
+	eps := make([]*discovery.EndpointSlice, 0, 2)
+	if enableIPv4 {
+		eps = append(eps, generateEps(discovery.AddressTypeIPv4, []string{"10.244.0.2"}))
+	}
+	if enableIPv6 {
+		eps = append(eps, generateEps(discovery.AddressTypeIPv6, []string{"fd00::100"}))
+	}
+
+	return svc, eps
+}
+
 func TestMetrics(t *testing.T) {
-	legacyregistry.Reset()
-	metrics.Register()
+	proxyConfig := antreaconfig.AntreaProxyConfig{
+		ProxyAll:             true,
+		ProxyLoadBalancerIPs: ptr.To(true),
+		ServiceProxyName:     testServiceProxyName,
+	}
+	originalEndpointSliceAPIAvailableFn := endpointSliceAPIAvailableFn
+	endpointSliceAPIAvailableFn = func(_ clientset.Interface) (bool, error) {
+		return true, nil
+	}
+	t.Cleanup(func() {
+		endpointSliceAPIAvailableFn = originalEndpointSliceAPIAvailableFn
+	})
 
-	for _, tc := range []struct {
-		name     string
-		isIPv6   bool
-		protocol binding.Protocol
+	testCases := []struct {
+		name                   string
+		proxierIPv4Enable      bool
+		proxierIPv6Enable      bool
+		svcIPv4Enabled         bool
+		svcIPv6Enabled         bool
+		expectedMetrics        map[string]int
+		expectedUpdatedMetrics map[string]int
```

Comment:
maybe add a comment to explain the difference between the 2 fields
I think we start with some initial resources and then we perform some updates?